lib/rucio/common/stomp_utils.py

```diff
@@ -15,145 +15,415 @@
+        """
+        if isinstance(fqdns, str):
+            fqdns = fqdns.split(',')
+
+        hosts_and_ports = []
+        for fqdn in fqdns:
+            try:
+                addrinfos = socket.getaddrinfo(fqdn.strip(), port, socket.AF_INET, 0, socket.IPPROTO_TCP)
+            except socket.gaierror as exc:
+                self._logger(logging.ERROR, "[broker] Cannot resolve domain name %s (%s)", fqdn.strip(), str(exc))
+                continue
+
+            hosts_and_ports.extend(addrinfo[4] for addrinfo in addrinfos)
+        if not hosts_and_ports:
+            self._logger(logging.WARNING, "[broker] No resolved brokers")
+        return hosts_and_ports
+
+    def _is_stalled(self, conn: Connection) -> bool:
+        """
+        Determine if a connection is stalled.
+
+        Args:
+            conn (Connection): The Connection object
+
+        Returns:
+            bool: Whether the connection has stalled.
+        """
+        received_heartbeat = getattr(conn, 'received_heartbeat', None)
+        if received_heartbeat is None or not any(self._config.heartbeats):
+            return False
+
+        heartbeat_period_seconds = max(0, self._config.heartbeats[0], self._config.heartbeats[1]) / 1000
+        if heartbeat_period_seconds == 0.:
+            return False
+
+        now = monotonic()
+        if received_heartbeat + 10 * heartbeat_period_seconds >= now:
+            return False
+
+        return True
+
+    def connect(self) -> Generator[Connection, None, None]:
+        """
+        Connect.
+
+        Yields:
+            Generator[Connection, None, None]: Each connection object after ensuring it's connected.
+        """
+        config = self._config
+        params = {'wait': True, "heartbeats": self._config.heartbeats}
+        self._logger(logging.WARNING, 'heartbeats: %s', self._config.heartbeats)
+        if config.use_ssl:
+            params.update(username=config.username, password=config.password)
+
+        for conn in self._conns:
+            if self._is_stalled(conn):
+                try:
+                    conn.disconnect()
+                except Exception:
+                    self._logger(logging.ERROR, "[broker] Stalled connection could not be disconnected")
             if not conn.is_connected():
-                conn.disconnect()
+                self._logger(logging.INFO, 'connecting to %s:%s', *conn.brokers[0])
+                # self._logger.info('connecting to %s', conn.transport._Transport__host_and_ports[0][0])
+                # if self._metrics is not None:
+                #     self._metrics.counter('reconnect.{host}').labels(host=conn.transport._Transport__host_and_ports[0][0].split('.')[0]).inc()
+                METRICS.counter('reconnect.{host}').labels(host=conn.brokers[0][0]).inc()
+                # METRICS.counter('reconnect.{host}').labels(host=conn.transport._Transport__host_and_ports[0][0].split('.')[0]).inc()
```

Comment:
Commented out code should be removed or a comment should be added to explain why we must keep it in production